pkg/apis/core/validation/shoot.go

```diff
@@ -93,6 +93,26 @@ func ValidateShootUpdate(newShoot, oldShoot *core.Shoot) field.ErrorList {
 	return allErrs
 }
 
+// ValidateShootTemplate validates a ShootTemplate.
+func ValidateShootTemplate(shootTemplate *core.ShootTemplate, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, metav1validation.ValidateLabels(shootTemplate.Labels, fldPath.Child("metadata", "labels"))...)
+	allErrs = append(allErrs, apivalidation.ValidateAnnotations(shootTemplate.Annotations, fldPath.Child("metadata", "annotations"))...)
+	allErrs = append(allErrs, ValidateShootSpec(shootTemplate.ObjectMeta, &shootTemplate.Spec, fldPath.Child("spec"))...)
+
+	return allErrs
+}
+
+// ValidateShootTemplateUpdate validates a ShootTemplate before an update.
+func ValidateShootTemplateUpdate(newShootTemplate, oldShootTemplate *core.ShootTemplate, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, ValidateShootSpecUpdate(&newShootTemplate.Spec, &oldShootTemplate.Spec, newShootTemplate.ObjectMeta, fldPath.Child("spec"))...)
+
+	return allErrs
+}
+
 // ValidateShootSpec validates the specification of a Shoot object.
 func ValidateShootSpec(meta metav1.ObjectMeta, spec *core.ShootSpec, fldPath *field.Path) field.ErrorList {
```

Comment:
Unlike other functions similar to this one (`ValidateSeedSpec`, `ValidateManagedSeedSpec`), this function does not take an `inTemplate bool` parameter. Instead, I see that some validation code was entirely removed. Shouldn't it be similar for all the three cases?

Reply:
This is exactly the discussion we had with @timebertt . Initially, it had the `inTemplate` parameter. Tim however argued, and I agreed with him, that the `inTemplate` parameter should be avoided if possible since it makes the validation code harder to maintain. To achieve this, we should distinguish clearly between the following two cases:

* Validation that is in contradiction with the `+optional` on a field and this is not justified by any special condition. In such cases, there is always defaulting code (or an admission plugin) to set the field to a valid value, however it only kicks in if an actual resource is created. Such validation should simply be removed. It should be removed carefully however to allow empty non-mandatory fields, but still reject non-empty fields if other conditions are violated. 
* Validation that really must be different between "in-template" and otherwise. We looked hard for ways to avoid the `inTemplate` parameter also for such cases but couldn't find a better approach.

So for `Shoot`, it turned out all validation code that failed the template validation fell in the first category, so an `inTemplate` parameter was not needed there. For `Seed` and `ManagedSeed` however we had cases in the second category, so the `inTemplate` parameter was kept.

Reply:
I agree that this parameter makes the whole validation code hard to maintain and that it should be avoided if possible. Great that you already discussed it and I think the outcome makes sense. Thanks for clarifying.